### PR TITLE
Auto-parse end2end tests

### PR DESCRIPTION
**Pull request by rfdavid**
An initial proposal to scan end-to-end files inside the e2e directory and dynamically run the tests.

This PR is related to the first task in the proposal TODO list:
- Set up the framework to auto-parse and run end2end tests without manually writing cpp code.
- Rewrite all read-only tests with new auto-parsing and running framework.

### How to add new end-to-end tests

For each e2e test, create a file named "test.group" containing the following fields:
```
-GROUP [group name]
-TEST [test name]
-DATASET [dataset]
-CHECK_ORDER true - use runTestAndCheckOrder() instead of runTest().
```

### Implementation details

- `e2e_read_test` recursively scan for `.test` files inside `test/test_files`
- If the directory contains the configuration file `test.group`, parse the file and extract the necessary information to create the tests dynamically
- Although `READ_ONLY` is in the test_files, it's been used only for informative purposes

### CODE REVIEW

File: test/runner/e2e_read_test.cpp
Comment (ray6080) [inline test/runner/e2e_read_test.cpp:40]: Can you take a look at `FileUtils::joinPath`? I think we can reuse that function to join path parts.

File: test/runner/e2e_read_test.cpp
Comment (ray6080) [inline test/runner/e2e_read_test.cpp:38]: I think you can leave "Test" postfix out of the `convertSnakeCaseToCamelCase` call, so it saves a bit processing, though doesn't matter much. 😄

**Review by ray6080 — CHANGES_REQUESTED** [review]
Nice start!

I have one design change that I hope to discuss with you.
Currently in this PR, we rely on the directory naming to figure out the test suite name and dataset path, which is fine, but limits the organization of test files, and coupling the directory name with test semantic is not a clean design in my mind. I think it's better to just have one file under the root level directory (`tinysnb`, `npy_1d`, etc.) that documents test suite name and dataset path.
You can see the first example in my proposal document. Let me know what do you think, we can discuss what's a better way.

**Review by ray6080 — COMMENTED** [review]
Also, I noticed one thing, when I actually run tests, I see message like this "TinysnbTest.Agg  # GetParam() = Agg". Do you have any ideas why we have `# GetParam() = Agg`?

**Comment by rfdavid** [comment]
> 

Thank you @ray6080 for the valuable feedback!
I totally agree, I'll work on parsing the `.group` file to extract the test suite properties (group, dataset, dbpath, others). I already have an idea in mind, I'll share it once I figure out some details.

**Comment by rfdavid** [comment]
> Also, I noticed one thing, when I actually run tests, I see message like this "TinysnbTest.Agg # GetParam() = Agg". Do you have any ideas why we have `# GetParam() = Agg`?

I realize I'm setting `value_param` inside `testing::RegisterTest` even though we're not using params. I'll remove it.

File: test/runner/e2e_read_test.cpp
Comment (rfdavid) [reply test/runner/e2e_read_test.cpp:11]: wdym by `move` both `testFiles` and `dataset`?

Comment (ray6080) [reply test/runner/e2e_read_test.cpp:11]: I mean you should use `std::move` here to avoid copy:
```
: testFiles(std::move(testFiles)), dataset(std::move(dataset)) {}
```

You don't need to add the "\_" postfix to class fields. We usually don't do that.
When there is shadowing of class variable names in function params, we usually add "_" postfix to function params.

File: test/runner/e2e_read_test.cpp
Comment (ray6080) [inline test/runner/e2e_read_test.cpp:39]: I'm not familiar with `recursive_directory_iterator`. Just one question here: I guess `recursive_directory_iterator` have the iteration order guarantee? It seems you rely on that as you are checking `FileUtils::getParentPathStem(entry) != previousDirectory`. Specifically, I just want to make sure that the following case is covered in this function.

```
test_files/test.group
test_files/t1.test
test_files/group1/test.group
test_files/group1/g1.test
test_files/group2/case1/test.group
test_files/group2/case1/g2c1.test
```

Alternatively, I think something like the following also makes sense, if you don't want to rely on the `recursive_directory_iterator`:
```
void scanTestFiles(const std::string& path, std::vector<TestConfig>& configs) {
    for (entry in path) {
        if (test group file exists) { 
            auto config = parseTestGroup(path); 
            if (config.isValid()) {
                configs.push_back(config);
            }
        }
        if (entry is directory) { scanTestFiles(entry.path, configs); }
    }
}

TestConfig parseTestGroup(const std::string& path) {
    auto testConfig = TestHelper::parseGroupFile(testGroupFile);
    for (entry in path) {
        if (entry is a file && entry's extension is ".test") {
            testConfig.files.push_back(entry.path().string());
        }
    }
    return testConfig;
}
```

Your call here. I don't have preferences over these two solutions.

**Review by ray6080 — APPROVED** [review]
LGTM. Please squash before you merge. Thanks!

File: test/runner/e2e_read_test.cpp
Comment (rfdavid) [reply test/runner/e2e_read_test.cpp:39]: Thanks, that's a good point.
From my tests, I haven't seen something like this (which will result in failure):
```
test_files/group2/case1/test.group
test_files/group1/g1.test
test_files/group2/case1/g2c1.test
```
However, looking at the documentation, it seems the order is not guaranteed. 
Your solution is more reliable and clear, I'll implement that.

Comment (rfdavid) [reply test/runner/e2e_read_test.cpp]: I initially implemented your code, the only necessary change was to add a bool to check if the current directory was processed and not create multiple configs. It worked, but I slightly changed it to check the directories first and then parse them. I believe it's cleaner, the functions are very small. I hope you're ok with that.

File: test/test_helper/test_helper.cpp
Comment (rfdavid) [reply test/test_helper/test_helper.cpp]: Added `setConfigValue` to solve this.
Once I start working on the test file parser, I was thinking of improving this parser by separating it into other classes and having better validation. I'll share more specifics soon.

File: test/runner/e2e_read_test.cpp
Comment (ray6080) [reply test/runner/e2e_read_test.cpp]: Sure! LGTM!

**Pull request merged**
